File: lenskart/frontend/src/types/product.ts
export interface Category {
  id: string
  name: string
  slug: string
  parentId?: string
  imageUrl?: string
  description?: string
  displayOrder: number
  isActive: boolean
  children?: Category[]
}

export type FrameType = 'FULL_RIM' | 'HALF_RIM' | 'RIMLESS'
export type FrameShape = 'ROUND' | 'SQUARE' | 'OVAL' | 'CAT_EYE' | 'RECTANGLE' | 'AVIATOR' | 'WAYFARER' | 'GEOMETRIC'
export type FrameMaterial = 'METAL' | 'PLASTIC' | 'TITANIUM' | 'TR90' | 'ACETATE' | 'WOOD'
export type Gender = 'MEN' | 'WOMEN' | 'UNISEX' | 'KIDS'
export type ViewType = 'FRONT' | 'SIDE' | 'ANGLE' | 'ON_FACE' | 'AR_OVERLAY'

export interface ProductImage {
  id: string
  productId: string
  variantId?: string
  url: string
  altText?: string
  displayOrder: number
  isPrimary: boolean
  viewType?: ViewType
}

export interface ProductVariant {
  id: string
  productId: string
  colorName: string
  colorHex?: string
  colorImageUrl?: string
  stockQty: number
  isAvailable: boolean
  images?: ProductImage[]
}

export type LensType = 'SINGLE_VISION' | 'BIFOCAL' | 'PROGRESSIVE' | 'ZERO_POWER'
export type LensCoating = 'ANTI_GLARE' | 'BLUE_CUT' | 'PHOTOCHROMIC' | 'TRANSITIONS' | 'NONE'
export type LensMaterial = 'STANDARD' | 'THIN' | 'ULTRA_THIN' | 'TRIVEX'

export interface LensOption {
  id: string
  type: LensType
  coating: LensCoating
  material: LensMaterial
  name: string
  description: string
  price: number
  isActive: boolean
  recommended?: boolean
}

export interface Product {
  id: string
  sku: string
  name: string
  slug: string
  description: string
  category: Category
  brand: string
  frameType: FrameType
  frameShape: FrameShape
  frameMaterial: FrameMaterial
  gender: Gender
  isPrescription: boolean
  isSunglasses: boolean
  isBlueLight: boolean
  basePrice: number
  discountPercent: number
  finalPrice: number
  rating: number
  reviewCount: number
  isActive: boolean
  isFeatured: boolean
  tags: string[]
  variants: ProductVariant[]
  images: ProductImage[]
  metaTitle?: string
  metaDescription?: string
  createdAt: string
  updatedAt: string
}

export interface ProductListResponse {
  content: Product[]
  totalElements: number
  totalPages: number
  currentPage: number
  pageSize: number
}

export interface ProductFilters {
  category?: string
  brand?: string[]
  minPrice?: number
  maxPrice?: number
  frameType?: FrameType[]
  frameShape?: FrameShape[]
  frameMaterial?: FrameMaterial[]
  gender?: Gender[]
  rating?: number
  isPrescription?: boolean
  isSunglasses?: boolean
  isBlueLight?: boolean
  tags?: string[]
  inStock?: boolean
}

export interface ProductSortOption {
  label: string
  value: string
  field: string
  direction: 'asc' | 'desc'
}

export interface Review {
  id: string
  productId: string
  userId: string
  userName: string
  userAvatar?: string
  rating: number
  title: string
  body: string
  images: string[]
  isVerifiedPurchase: boolean
  helpfulCount: number
  isApproved: boolean
  createdAt: string
}

export interface ReviewStats {
  average: number
  total: number
  distribution: Record<number, number>
}

File: lenskart/backend/app/schemas/product.py
from pydantic import BaseModel, ConfigDict, Field
from typing import Optional, List
from decimal import Decimal
import uuid
from datetime import datetime


class CategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    slug: str
    description: Optional[str] = None
    imageUrl: Optional[str] = Field(None, validation_alias='image_url')
    displayOrder: int = Field(validation_alias='display_order')
    isActive: bool = Field(validation_alias='is_active')


class VariantOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    colorName: str = Field(validation_alias='color')
    colorHex: Optional[str] = Field(None, validation_alias='color_hex')
    stockQty: int = Field(validation_alias='stock_qty')
    priceAdjustment: Decimal = Field(validation_alias='price_adjustment')
    isDefault: bool = Field(validation_alias='is_default')


class ImageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    url: str
    altText: Optional[str] = Field(None, validation_alias='alt_text')
    displayOrder: int = Field(validation_alias='display_order')
    isPrimary: bool = Field(validation_alias='is_primary')


class LensOptionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    description: Optional[str] = None
    type: str
    price: Decimal
    features: Optional[str] = None
    isRecommended: bool = Field(validation_alias='is_recommended')


class ProductOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    slug: str
    brand: Optional[str] = None
    description: Optional[str] = None
    categoryId: Optional[uuid.UUID] = Field(None, validation_alias='category_id')
    categoryName: Optional[str] = None
    basePrice: Decimal = Field(validation_alias='price')
    finalPrice: Decimal = Field(Decimal('0'), validation_alias='final_price')
    discountPercent: int = Field(validation_alias='discount_percent')
    frameType: Optional[str] = Field(None, validation_alias='frame_type')
    frameShape: Optional[str] = Field(None, validation_alias='frame_shape')
    frameMaterial: Optional[str] = Field(None, validation_alias='frame_material')
    gender: Optional[str] = None
    isFeatured: bool = Field(validation_alias='is_featured')
    isNewArrival: bool = Field(validation_alias='is_new_arrival')
    isBestseller: bool = Field(validation_alias='is_bestseller')
    isBlueLight: bool = Field(validation_alias='has_blue_light')
    rating: Decimal = Field(validation_alias='average_rating')
    reviewCount: int = Field(validation_alias='review_count')
    variants: List[VariantOut] = []
    images: List[ImageOut] = []
    createdAt: datetime = Field(validation_alias='created_at')


class ReviewOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    userName: str = Field(validation_alias='user_name')
    rating: int
    title: Optional[str] = None
    body: Optional[str] = None
    isVerifiedPurchase: bool = Field(validation_alias='is_verified_purchase')
    helpfulCount: int = Field(validation_alias='helpful_count')
    createdAt: datetime = Field(validation_alias='created_at')


class ReviewIn(BaseModel):
    product_id: uuid.UUID
    rating: int
    title: Optional[str] = None
    body: Optional[str] = None
